### PR TITLE
[WIP] Haiku support

### PR DESCRIPTION
**Pull request by qyot27**
What works: compiling the library into libavisynth.so
What doesn't: actually using it in either avs2yuv or FFmpeg

Haiku is a re-implementation of BeOS, and something like this was brought up in a (long-since-closed) issue on Haiku's bug tracker, way back when expecting the Ruby-based 3.0 was still a thing:
https://dev.haiku-os.org/ticket/3907

I've recently been tinkering on getting AviSynth+ running on Haiku, but I've hit a bit of a roadblock.  The library compiles, but when I attempt to load it in other applications (avs2yuv, or an FFmpeg that's been rebuilt with *--enable-avisynth*), I'm getting errors related to string handling.  But the backtraces from gdb don't seem to point at a particular area and the message given by the system about the error doesn't really match the parts of the code that use std::string.

The relevant bits of error and debug messages are below:
```
terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::erase: __pos (which is 18446744073709551615) > this->size() (which is 1)
Abort
Kill Thread
```

gdb output from avs2yuv
```
Program received signal SIGTRAP, Trace/breakpoint trap.
0x0000014232aef1c1 in _kern_debugger () from /boot/system/lib/libroot.so
(gdb) bt
#0  0x0000014232aef1c1 in _kern_debugger () from /boot/system/lib/libroot.so
#1  0x0000014232ae6c98 in abort () from /boot/system/lib/libroot.so
#2  0x0000014232ae7077 in __gnu_cxx::__verbose_terminate_handler ()
   from /boot/system/lib/libroot.so
#3  0x0000014232b7d369 in __cxxabiv1::__terminate ()
   from /boot/system/lib/libroot.so
#4  0x0000014232b7d3b1 in std::terminate () from /boot/system/lib/libroot.so
#5  0x0000014232b7d50c in __cxa_throw () from /boot/system/lib/libroot.so
#6  0x000001b95e921cab in std::__throw_out_of_range_fmt ()
   from /boot/system/lib/libstdc++.so.6.0.25
#7  0x000001f8026536d6 in std::string::_M_check ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#8  0x000001f80266393f in std::string::erase ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#9  0x000001f80265fdf5 in PluginManager::AddAutoloadDir ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#10 0x000001f80267bd93 in ScriptEnvironment::ScriptEnvironment ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#11 0x000001f802686eec in CreateScriptEnvironment2 ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#12 0x000001f802686e8c in CreateScriptEnvironment ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
#13 0x000001f8026b12b2 in avs_create_script_environment ()
   from /boot/home/avsplus_build/lib/libavisynth.so.3.6.2
---Type <return> to continue, or q <return> to quit---
#14 0x00000043ffa870a8 in main ()
(gdb) 
```

### CODE REVIEW

**Comment by pylorak** [comment]
> but I've hit a bit of a roadblock
> ... the backtraces from gdb don't seem to point at a particular area 
The trace says it happens in PluginManager::AddAutoloadDir() while calling string:erase(), and the console error reaffirms that (or at least that the exception is happening in a call to string::erase)

The only such call in AddAutoloadDir is on line 578 (`ExeFileDir.erase`). At this point ExeFileDir should more-or-less still be equal to ExeFilePath (barring the \ -> / replacement) which is the path to the running executable, and the console error says its size is 1, which is obviously wrong. The complete path to our executable shouldn't consist of a single char. (The erase then seems to throw because ExeFileDir.rfind('/') returns string::npos which is -1, which is a very large number when interpreted as the erase arg size_t, which is larger than our string size of 1. But this is not the interesting part, the real question is why ExeFilePath is wrong.)

Long story short, I suspect one of those readlinks taken above is wrong for Haiku, and you'll need to adjust and specialize the code how you get the running executable under Haiku.

**Comment by qyot27** [comment]
A very similar thing happened when macOS and BSD were added, because the `_NSGetExecutablePath` and 'go ahead and just make the BSDs that do support `/proc` use that for right now' assumptions weren't made and were left blank.  But the odd part is that in that situation, the library still *functioned*, it just wasn't able to automatically add the autoload dirs and a user had to use LoadPlugin to get around it.  Unless the issue is that Haiku's different filesystem structure is what's at play here.  It should be easy enough to test, at least.

**Comment by qyot27** [comment]
And I've just tested it; it's definitely not a long-term solution, but hiding line 578 in PluginManager.cpp behind an `#ifndef AVS_HAIKU` allows the library to actually function in avs2yuv and FFmpeg with the command line I was using:
```
LIBRARY_PATH=$LIBRARY_PATH:$HOME/avsplus_build/lib ./avs2yuv test.avs -o test.y4m
```
or
```
LIBRARY_PATH=$LIBRARY_PATH:$HOME/avsplus_build/lib ./ffmpeg -i test.avs
```
The simple `Version()` (or `Version().ConvertToYUV420()` in avs2yuv) script was able to be encoded to y4m or its properties detected in the format probing functions.  The resultant y4m played back correctly, so it was actually doing what it was supposed to.

But knowing that this is the only real issue hindering Haiku support is pretty awesome.

It almost certainly needs a Haiku-specific solution to be written, along with the sections in avisynth.cpp that detect physical and available memory (which I think can be accessed via their kernel/OS.h header, but I'm not entirely certain about that).  That's currently hardcoded to a max of 1GB physical memory for Haiku just to make sure it wouldn't cause problems, and there's nothing but a to-do note in the available pages detection.

**Comment by qyot27** [comment]
Okay, fixed up a little and force pushed, so the branch as it exists does build and run on Haiku in the basic tests I was running.  But as I note in the commit message, the physical/available memory functions are either hardcoded to 1GB or absent, and the executable path needs its own native solution (it probably means users will need to LoadPlugin anything they want to use until that gets resolved, hopefully soon).

**Comment by pylorak** [comment]
As for getting the amount of installed and free memory, `get_system_info()` from `kernel/OS.h` should do the job. Haiku seems to use 4K pages, so the total amount of installed memory should be 4096 x `max_pages` returned by `get_system_info`. The amount of free memory could be `max_pages` - `min_pages`, or `free_memory` directly, not sure if there's a difference. 

No idea about getting the app path. Doesn't poking around in /proc reveal something usable? I don't have a Haiku VM set up to have a quick look at this.

**Comment by qyot27** [comment]
Since Haiku isn't a Unix-like OS, it doesn't have /proc; and for that matter, Linux is probably the most prominent one in that family that seems to want to keep it around - the *BSDs (including OSX/Darwin, although their alternative was easy enough to plug in) have moved to either considering it deprecated and not mounting it by default, or outright removing it.

I'd figured the `*_pages` entries in `get_system_info()` were what I needed, but when I tried to get the values out of there, I was forgetting pieces or was doing things wrong enough (not enough sleep or not enough recent practice dealing with structs, I dunno) that I was getting compilation errors and need to go back and try again.

**Comment by qyot27** [comment]
>and the executable path needs its own native solution (it probably means users will need to LoadPlugin anything they want to use until that gets resolved, hopefully soon)

Yeah, turns out this is wrong; the autoload dirs still function on Haiku, as I built FFMS2 (didn't require any patching, either), symlinked it into the default autoload directory, and didn't get any complaints over the script loading.  The video loaded just fine.

Which would leave it as possibly affecting AddAutoloadDir, but in avisynth.cpp, where the autoload directory behavior is handled for non-Windows, this is the relevant block:
```
#else
    // system_avs_plugindir relies on install path, it gets
    // defined in avisynth_conf.h.in when configuring.

    std::string user_avs_plugindir = std::getenv("HOME");
    std::string user_avs_dirname = "/.avisynth";
    user_avs_plugindir.append(user_avs_dirname);

    plugin_manager->AddAutoloadDir(user_avs_plugindir, false);
    plugin_manager->AddAutoloadDir(system_avs_plugindir, false);
#endif
```

And it still worked.  So is the use of `ExeFileDir.erase` meant to clean up the strings that are being passed around and hiding the line may 'succeed' in allowing it to work but potentially cause a resource leak?  Or is that line failing maybe the result of a conflict of some sort intrinsic to Haiku that might render that line unnecessary, hence why hiding it seems to be okay?

**Comment by qyot27** [comment]
And I've also posted about the topic on the Haiku Community forum:
https://discuss.haiku-os.org/t/avisynth-scripting-nle-frameserver/10086

Which should be another way of keeping abreast of anything specifically Haiku-related.

**Comment by pylorak** [comment]
> And it still worked. So is the use of `ExeFileDir.erase` meant to clean up the strings that are being passed around and hiding the line may 'succeed' in allowing it to work but potentially cause a resource leak? Or is that line failing maybe the result of a conflict of some sort intrinsic to Haiku that might render that line unnecessary, hence why hiding it seems to be okay?

Actually, neither of those. The `ExeFileDir` (and thus `ExeFilePath`) are needed to support the `PROGRAMDIR` variable in Avisynth scripts. So only in your specific test case the line wasn't needed which is why your autoload test worked, because the variable wasn't used. But if you tried a script that uses `PROGRAMDIR` as a parameter to `AddAutoloadDir`, that script's autoload would be broken without line 578.

> I'd figured the *_pages entries in get_system_info() were what I needed, but when I tried to get the values out of there, I was forgetting pieces or was doing things wrong enough (not enough sleep or not enough recent practice dealing with structs, I dunno) that I was getting compilation errors and need to go back and try again.

Rough outline:
```
system_info sysinf;
get_system_info(&sysinf);
uint64_t total_mem = 4096 * sysinf.max_pages;
uint64_t free_mem = sysinf.free_memory;
```

I'm not sure about this, but I think using `4096 * (sysinf.max_pages - sysinf.used_pages)` would give us the available virtual memory instead of the available physical memory, so I went for `free_memory` instead. This needs to be verified for correctness.

**Comment by extrowerk** [comment]
For the exec path refer: https://git.haiku-os.org/haiku/commit/?id=107712bf01c9777592fd24f7c06c455d1110df0a

**Comment by pylorak** [comment]
> For the exec path refer: https://git.haiku-os.org/haiku/commit/?id=107712bf01c9777592fd24f7c06c455d1110df0a

Hi extrowerk,

Thank you for the pointers. Unfortunately we have zero Haiku/BeOS know-how and so we're (or at least I am) missing some basic BeOS-specific knowledge to make sense of that code you linked.

Some questions:

1. The few removed lines right at the top is what we need, right? They seem to be the right code but they are the ones being removed in the patch so this makes me insecure if those are what's recommended.

```
app_info info;
be_app->GetAppInfo(&info);
// then inspect info.directory
```

2. Assuming that's the way above to go, what the hell is `be_app`? Even when I look at the whole file, it isn't initialized or even declared anywhere. So it must come from a header. Which header? Are there some other external functions we need to call before we cann access `be_app` at all? Is that some application-specific variable or a standard BeOS-thing? etc.

3. When we finally have our app_info, how do we get our directory? The member `name` returns the "leaf entry" according to the docs, whatever that is supposed to mean. I guess the leaf directory entry, so that'll only be the filename without the path. Meaning we need the `directory` member instead, but that returns the inode number and not the path. So then how do we get the directory path (as a string) from the inode number under BeOS?

**Comment by qyot27** [comment]
`be_app` appears to be part of the [BApplication class](https://www.haiku-os.org/docs/api/classBApplication.html), provided by the [Application.h header](https://www.haiku-os.org/docs/api/Application_8h.html).  [Related section in the Be Book](https://www.haiku-os.org/legacy-docs/bebook/BApplication_Overview.html).

I'm not sure if the fact AviSynth+ is a shared library loaded by a client program and not a standalone application changes the calculus for this at all.  This is uncharted waters for me too.

As regards the mempages blocks, I'm not really sure what I should be looking for to see if it's failing vs. doing its job correctly, because watching the Activity Monitor or the output of `top` or the processing fps in FFmpeg or the generic memory reporting in the About Haiku window seem to act more or less the same whether the blocks are changed as suggested above or if they're left the way I had them.  But I fully admit that may be because of the VM I'm using to test it; maybe I'm giving the VM too much RAM to be able to see a clear effect of the available memory and total memory being able to be gleaned correctly vs. just being told it has X and leaving the available pages part blank (but then seemingly just blowing past the capped 1G total memory anyway depending on what I tell the script to do).

File: avs_core/CMakeLists.txt [hunk 1]
```diff
@@ -102,10 +102,16 @@ if (CMAKE_COMPILER_IS_GNUCXX AND CMAKE_CXX_COMPILER_VERSION VERSION_LESS 9)
...
 
-# Windows DLL dependencies
+# Per-OS link dependencies
 if (MSVC OR MINGW)
   target_link_libraries("AvsCore" "uuid" "winmm" "vfw32" "msacm32" "gdi32" "user32" "advapi32" "ole32" "imagehlp")
-else()
+endif()
+
+if (HAIKU)
+  target_link_libraries("AvsCore" "pthread" "${FSLIB}")
```
Comment (CodeforEvolution) [inline avs_core/CMakeLists.txt:111]: Hello! Just a quick note: 
Haiku's libpthread is only a "fake library" for compatibility purposes, and instead, you'll want to link to libroot. In other words, you'll want to replace "pthread" with "root" on this line.

Comment (qyot27) [reply avs_core/CMakeLists.txt]: I knew it was true for dlopen, that's why that one was omitted.  I don't remember if I'd seen that pthreads were also counted in that too, but thanks for confirming it.

Comment (CodeforEvolution) [reply avs_core/CMakeLists.txt]: No problem!

**Comment by qyot27** [comment]
Merged in as part of 3.7.

The missing exec_path stuff is still not there, but that's partially due to my not really knowing how useful it really is on non-Windows anyway; the variable that AviSynth+ could use that would require it to actually work is not related to autoloading, but to finding the system install area in lieu of autoloading being turned off, and the user not wanting to explicitly specify the system path.  It's probably fairly marginal of a concern on either *nix or Haiku.  Of course, eventually it might be/probably will be fully fleshed out, but I've been feeling a bit of pressure of just sitting on this, and the static library and PPC stuff.
